torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -377,6 +378,318 @@ bool IsBlockReturnTypeSame(Node* n) {
+  for (auto i = minus_one_pos + 1; i < reshape_size; i++) {
+    int64_t cur_shape = reshape[i] == 0
+        ? input_shape[i + input_shape_size - reshape_size]
+        : reshape[i];
+    final_shape.push_back(cur_shape);
+  }
+  return final_shape;
+}
+
+void UpdateRank(Value* value, size_t rank) {
+  ConstantValueMap::SetRank(value->debugName(), rank);
+  c10::optional<size_t> rank_opt = rank;
+  auto sym_shape = ::c10::SymbolicShape(rank_opt);
+  value->setType(
+      value->type()->cast<TensorType>()->withSymbolicShapes(sym_shape));
+}
+
+void UpdateShape(Value* value, const std::vector<int64_t>& shape_size) {
+  if (shape_size.empty()) {
+    UpdateRank(value, 0);
+    return;
+  }
+  auto vary_shape_size = c10::VaryingShape<int64_t>(shape_size);
+  ConstantValueMap::SetShape(value->debugName(), vary_shape_size);
+  value->setType(value->type()->cast<TensorType>()->withSymbolicShapes(
+      ::c10::SymbolicShape(shape_size)));
+}
+
+// As an addition to onnx shape inference, this function leverages constant
+// folding and a per-Op based process to update rank/shape for the graph, also
+// it update ConstantValueMap accordingly.
+void ComputeConstant(Node* n, int opset_version) {
+  if (n->kind() == ::c10::onnx::Constant) {
+    if (n->kindOf(attr::value) == AttributeKind::t) {
+      at::Tensor const_val = n->t(attr::value);
+      at::Tensor const_val_copy =
+          at::empty(const_val.sizes(), const_val.options());
+      const_val_copy.copy_(const_val);
+      ConstantValueMap::SetValue(n->output()->debugName(), const_val_copy);
+    }
+    return;
+  }
+  auto only_rank_available = false;
+  size_t rank = 0;
+
+  // Constant folding.
+  auto const_fold_val = ComputeConstantFolding(n, opset_version);
+  if (const_fold_val.has_value()) {
+    at::Tensor const_fold_val_copy = at::empty(
+        const_fold_val.value().sizes(), const_fold_val.value().options());
+    const_fold_val_copy.copy_(const_fold_val.value());
+    ConstantValueMap::SetValue(n->output()->debugName(), const_fold_val_copy);
+    UpdateShape(n->output(), const_fold_val_copy.sizes().vec());
+  }
+
+  switch (n->kind()) {
+    case ::c10::onnx::Shape: {
+      if (ConstantValueMap::HasShape(n->input()->debugName())) {
+        auto shape_size = ConstantValueMap::GetShape(n->input()->debugName())
+                              .concrete_sizes();
+        if (shape_size.has_value()) {
+          std::vector<int64_t> shape_value;
+          for (const auto& v : shape_size.value()) {
+            shape_value.emplace_back(static_cast<int64_t>(v));
+          }
+          // TODO: getDevice() ?
+          auto options = c10::TensorOptions().dtype(at::kLong).device(at::kCPU);
+          auto shape_value_size = static_cast<int64_t>(shape_value.size());
+          auto f =
+              at::from_blob(shape_value.data(), {shape_value_size}, at::kLong)
+                  .to(at::kCPU);
+          // Need copy here
```

Comment:
could you elaborate why copy is needed?

Reply:
Looking at definition of `from_blob`, there are several overloading, and see description here in https://github.com/pytorch/pytorch/blob/master/tools/autograd/templates/variable_factories.h#L54, `from_blob` exposes the given `data` as a `Tensor` without taking ownership of the original data. So when this variable `shape_value` goes out of scope, the underlying data (inserted in the hashmap) can be corrupted.